docs: #3714 write Bug/Feedback submission tech spec

goodhoko opened the pull request:
cc @tsusanka @tomexx 

Rendered version: https://github.com/trezor/trezor-suite/blob/7725d0715bfb7f575001a3a4f05542c3e8a4be24/docs/specs/bug_and_feedback_submission.md

Not sure who else (if any) should I tag.

goodhoko commented:
@tsusanka I posted this as PR in Suite mainly for PR reviewing tooling. This doesn't necessarily need to be merged -- we can keep the final version somewhere else. Though, I think the Suite repo isn't a bad place for tech specs at all. .)

tsusanka commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -0,0 +1,56 @@
+_note: Though we're dealing with two distinct streams here (bugs and feedback) the technical infrastructure will be mostly identical. Hence, in this spec, we'll call bugs and feedback collectively as messages._
+
+## Scope
+At on side we have the Suite application that can send HTTP requests. On the other side we use [Keboola](https://www.keboola.com/) as the central hub for all data collected around Trezor. This spec describes how we connect these two sides.
```
At on side -> At one side

goodhoko replied:
Addressed in a51604d2c38c209d80e8f69e308476f44e92d728

Resolving.

tsusanka commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -0,0 +1,56 @@
+1. Messages contain mostly structured textual data and sometimes can contain an image.
+2. Messages need to be consumed by Keboola.
+3. The submitted messages must be kept private as they might contain sensitive user information. Only after going thru a manual triage they might end up in a publicly accessible place like, for example, GitHub issues.
+4. The messages should be enriched with a timestamp and client's IP on the server side for client's clock might be off and determining public IP on client side requires an external service anyway.
```
Why IP? I understand AWS has it but I would prefer not to send it further.

goodhoko replied:
Good question. The IP indeed isn't needed for the sole purpose of collecting bugs or feedback.

When I'm thinking about it now: IP *might* become useful in case someone will decide to flood our 'API' with fake bug reports as common identifier of those fake reports. Still, this is quite hypothetical scenario and I personally would prefer to not collect IPs as well.

tsusanka replied:
Yes, let's not do it until we have a clear and justifiable reason to do so.

goodhoko replied:
Removed in 0bd3efc2761d8ff96dca6c0ef57764c4ac3059b8

tsusanka commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -0,0 +1,56 @@
+- `/suite/bugs/devel`
+- `/suite/bugs/prod`
+
+with the same exact logging configuration. Tomas ferko will re-use the python scripts in Keboola. 
```
We could ping Tom, he should review it anyway.

goodhoko replied:
Will do. Resolving.

tsusanka commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -0,0 +1,56 @@
+
+with the same exact logging configuration. Tomas ferko will re-use the python scripts in Keboola. 
+
+If user includes an image in a massage it will be uploaded separately and the message will contains just a textual reference to the image. We will provision a pair (for production and dev/staging) of S3 buckets to store the images along with an IAM user who can [only `putObject`](https://stackoverflow.com/questions/37739086/how-to-allow-only-putobject-permissions-on-specific-directory-in-amazon-s ...
```
Nits:
- Massage :)
- message will contain (no s)

goodhoko replied:
> Massage :)

lol .D

Fixed both in 491aa0c7a83c6c92d945801442943e0e349220d5

Thanks. Resolving.

tsusanka reviewed (commented):
Btw should we add an option to the messaging system that is capable of turning off this feature? :thinking: I am just considering the "DDoS" attack of S3. We might want to turn off this feature then.

goodhoko commented:
> Btw should we add an option to the messaging system that is capable of turning off this feature? 🤔 I am just considering the "DDoS" attack of S3. We might want to turn off this feature then.

Turning the feature off in Suite via messaging system could prevent large amounts of *real* bug reports. I argue that is something we never want to do. If we'll get shitloads of reports sent from the app we want to keep them and pay for the storage. Disabling the feature in suite wouldn't prevent anyone from flooding the api directly. That would have to be done server-side. IE I don't think this is necessary. .)

[I added](https://github.com/trezor/trezor-suite/pull/3764/commits/551730a60cb1aa68ddd8327b6c7eeba8a4420dcd) two new sections to this spec: one that exactly describe what data we'll send with the messages and one that outlines (very briefly) what are the next steps for this feature. Plese check them out. @tsusanka

tsusanka commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -48,9 +48,40 @@ When putting an object into a bucket [the client must give it a key](https://doc
+### Feedback
+- `rating` 1-5, 1 is best, 5 is worst.
+- `description` (optional) User's impressions of the app.
+- `screenshot` (optional) UUID of linked image.
```
I know it is in design but does it make sense to attach a screenshot to a _Feedback_? Could you discuss this with @matejzak?

glacierbits replied:
It's not the core functionality of the feedback part, but we'd still like to support it: The user can indicate in a simple drag&drop way what they are talking about (their specific use case).

tsusanka replied:
Well okay then :)

tsusanka commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -48,9 +48,40 @@ When putting an object into a bucket [the client must give it a key](https://doc
+- `screenshot` (optional) UUID of linked image.
+
+### Common
+The fields sent with both bugs and feedback. All device-related fields are optional because the can be no device connected. Most of them can be reused from the [GitHub Issue feature](https://github.com/trezor/trezor-suite/blob/78736782fbd6556f536fcbc989ba15cf0b367555/packages/suite/src/services/github.ts#L36).
```
the -> there

But btw that is not true because you can't access the app until there is a Trezor connected (until we do #3729). Or will one be able to open the right Learn & Discover menu also on the Connect your Trezor page?

goodhoko replied:
> Or will one be able to open the right Learn & Discover menu also on the Connect your Trezor page?

I don't think so. cc @matejzak 

Removed the optionality in 76c10cd44f75cea6a017e0bd790508710074b0fe.

glacierbits replied:
We'd like to support the option to access App & Coin Settings without connected device in the next sprint #3729, including the Learn & Discover.

tsusanka commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -48,9 +48,40 @@ When putting an object into a bucket [the client must give it a key](https://doc
+- `device_type` (optional) eg. `T` or `One`
+- `firmware_version` (optional) versions
+- `firmware_revision` (optional) commit sha of the build of FW
+- `firmware-type` eg. `bitcoin-only` or `regular`
```
Maybe it is a crazy idea but I was considering to have a field called _markdown_ which would contain these fields and as well the bug description (again inspiration [here](https://github.com/trezor/trezor-suite/blob/78736782fbd6556f536fcbc989ba15cf0b367555/packages/suite/src/services/github.ts#L36)). This way Alex could move the bug report from the Google Sheet (or wherever) really simply by just copy & pasting this field to Github.

goodhoko replied:
I'd rather do that in Keboola or whichever tool the data ends up in. Mainly to reduce duplication and to not transmit and store more data than necessary.

OTOH the issue template is already present in Suite's code so we might just easily reuse it there. [Added](7725d0715bfb7f575001a3a4f05542c3e8a4be24) the field to the Bug fields as it IMO doesn't make much sense to have the template in feedback fields.

tsusanka commented:
> Turning the feature off in Suite via messaging system could prevent large amounts of real bug reports. I argue that is something we never want to do. If we'll get shitloads of reports sent from the app we want to keep them and pay for the storage. Disabling the feature in suite wouldn't prevent anyone from flooding the api directly. That would have to be done server-side. IE I don't think this is necessary. .)

That is a strange argument. Imagine DDoS with 100 000 reports. How does it help that you'll have 100 real reports there? I do not think will be able to triage these anyway. Take it as a last possible resort solution, something we'll hopefully won't use.

But I guess we can nuke the infrastructure in the worst case (because it should not happen hopefully). Let's just make sure we provide some nice error message if the request fails.

goodhoko commented:
> That is a strange argument. Imagine DDoS with 100 000 reports. How does it help that you'll have 100 real reports there? I do not think will be able to triage these anyway. Take it as a last possible resort solution, something we'll hopefully won't use.

Sorry, I didn't write that particularly clear. I wanted to stress that _turning the feature off via the messaging system wouldn't prevent 'DDOSing' the backend in anyway_. It can prevent submitting messages/images via the Suite app which is essentially just a GUI but the API remains accessible.

It's safe to assume an attacker attempting this attack would use the API directly instead of trying to submit large amounts of data by clicking around the app. It then follows that messages submitted directly from the app can be generally considered as genuine (except someone trolling us).

Note that this is something we can't entirely prevent. We are deploying an API that must accept requests without authentication because there's no user to authenticate and the client code is distributed in a way that makes it very easy to extract any credentials we'd try to package with it. As far as I know there really isn't much we can do to prevent this kind of attack that wouldn't affect UX in a significant way. Except the mentioned IP-based rate-limiting/filtering.

> Let's just make sure we provide some nice error message if the request fails.

Sure, let's keep an eye on it in @tomexx's coming PRs implementing the UI. .)

tsusanka commented:
Ah I understand. All good then 👍

goodhoko commented on docs/specs/bug_and_feedback_submission.md:
```diff
@@ -0,0 +1,88 @@
+
+### Bugs
+- `description` The textual description of the bug written by the user.
+- `path` Current url path. Corresponds to the "category" in the Designs.
```
This field awaits a clarification in the product spec. (There's one open comment I'd like to link but Notion doesn't provide permalinks for comments. ./ )

goodhoko replied:
Matej specified a static list of few categories. We'll send these as a string. In the app we'll write a function that will map routes to categories to support the `current screen` feature in the bug submission.

This spec edited [here](https://github.com/trezor/trezor-suite/pull/3764/commits/6a4ee408c230c7211083977eb51d17475f7df279).

Resolving.

tomexx reviewed (approved):
Nothing to add really :). Good job ;)

goodhoko commented:
@tsusanka I rebased and squashed all the commits. This is ready to be merged.

The next question is who will provision the required AWS infrastructure? .)

tsusanka commented:
> The next question is who will provision the required AWS infrastructure? .)

I will take care of it. Let's merge this and I'll try to do it this week.

goodhoko commented:
@tsusanka Thanks! Don't hesitate to ask if there's anything unclear in the spec.
